jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/ResponseNotifier.java

```diff
@@ -100,54 +99,260 @@ private void notifyHeaders(Response.HeadersListener listener, Response response)
+                    return null;
+                }
+
+                Content.Chunk result = chunk;
+                if (result != null && !result.isTerminal())
+                    chunk = AlreadyReadChunk.INSTANCE;
+                if (LOG.isDebugEnabled())
+                    LOG.debug("Content source #{} reading current chunk {}", index, result);
+                return result;
+            }
+
+            @Override
+            public void demand(Runnable demandCallback)
+            {
+                if (this.demandCallback != null)
+                    throw new IllegalStateException();
+                this.demandCallback = Objects.requireNonNull(demandCallback);
+                Content.Chunk currentChunk = this.chunk;
+                if (LOG.isDebugEnabled())
+                    LOG.debug("Content source #{} demand while current chunk is {}", index, currentChunk);
+                if (currentChunk == null || currentChunk == AlreadyReadChunk.INSTANCE)
+                    registerDemand();
+                else
+                    onDemandCallback();
+            }
+
+            @Override
+            public void fail(Throwable failure)
+            {
+                Content.Chunk currentChunk = chunk;
+                if (LOG.isDebugEnabled())
+                    LOG.debug("Content source #{} fail while current chunk is {}", index, currentChunk);
+                if (currentChunk instanceof Content.Chunk.Error)
+                    return;
+                if (currentChunk != null)
+                    currentChunk.release();
+                this.chunk = Content.Chunk.from(failure);
+                onDemandCallback();
+                registerFailure(failure);
+            }
+        }
+
+        private static class AlreadyReadChunk implements Content.Chunk
+        {
+            static final AlreadyReadChunk INSTANCE = new AlreadyReadChunk();
+
+            @Override
+            public ByteBuffer getByteBuffer()
+            {
+                throw new UnsupportedOperationException();
+            }
+
+            @Override
+            public boolean isLast()
+            {
+                throw new UnsupportedOperationException();
+            }
+
+            @Override
+            public void retain()
+            {
+                throw new UnsupportedOperationException();
+            }
+
+            @Override
+            public boolean release()
+            {
+                throw new UnsupportedOperationException();
+            }
+
+            @Override
+            public String toString()
+            {
+                return getClass().getSimpleName();
+            }
+        }
+    }
+
+    private void driveDemandLoop(Content.Source contentSource)
+    {
+        Content.Chunk read = contentSource.read();
+        if (read != null)
+            read.release();
+        if (read == null || !read.isLast())
+            contentSource.demand(() -> driveDemandLoop(contentSource));
```

Comment:
You mean `Content.Source.consumeAll()`? You're right, there is a slight difference that `driveDemandLoop()` does not loop with `while (true)` but that makes no difference here since no listener is being called. 
Let me get rid of `driveDemandLoop()`.